src/Microsoft.Data/BaseColumn.cs

```diff
@@ -0,0 +1,347 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Data
+{
+    /// <summary>
+    /// The base column type. All APIs should be defined here first
+    /// </summary>
+    public abstract partial class BaseColumn
+    {
+        public BaseColumn(string name, long length, Type type)
+        {
+            Length = length;
+            Name = name;
+            DataType = type;
+        }
+
+        private long _length;
+        public long Length
+        {
+            get => _length;
+            protected set
+            {
+                if (value < 0)
+                    throw new ArgumentOutOfRangeException();
+                _length = value;
+            }
+        }
+
+        public long NullCount { get; protected set; }
+
+        public string Name { get; set; }
+
+        public Type DataType { get; }
+
+        public object this[long rowIndex]
+        {
+            get => GetValue(rowIndex);
+            set => SetValue(rowIndex, value);
+        }
+
+        protected virtual object GetValue(long rowIndex) => throw new NotImplementedException();
+        protected virtual object GetValue(long startIndex, int length) => throw new NotImplementedException();
+
+        protected virtual void SetValue(long rowIndex, object value) => throw new NotImplementedException();
+
+        public object this[long startIndex, int length]
+        {
+            get => GetValue(startIndex, length);
```

Comment:
Just as a note: On a notebook, most calls such as the following will still box at the moment:
`
var intColumn = df.Column('intColumn'); //intColumn is a PrimitiveColumn<int>, but is returned as a BaseColumn
var thisWillBox = intColumn[2];
`